File: LEETCODE_498_Diagonal_Traverse.java
import java.util.*;

public class LEETCODE_498_Diagonal_Traverse {

    private void reverse(List<Integer> arr) {
        int start = 0, end = arr.size()-1;
        while(start <= end) {
            int temp = arr.get(start);
            arr.set(start++, arr.get(end));
            arr.set(end--, temp);
        }
    }

    private List<Integer> traverse1(int row, int[][] mat) {
        List<Integer> result = new ArrayList<>();
        for(int i = row, j = 0; i >= 0 && j < mat[0].length; i--, j++) {
            result.add(mat[i][j]);
        }
        return result;
    }

    private List<Integer> traverse2(int col, int[][] mat) {
        List<Integer> result = new ArrayList<>();
        for(int i = mat.length-1, j = col; i >= 0 && j < mat[0].length; i--, j++) {
            result.add(mat[i][j]);
        }
        return result;
    }

    public int[] findDiagonalOrder(int[][] mat) {
        boolean reverse = false;
        int[] answer = new int[mat.length * mat[0].length];
        int index = 0;
        for(int i = 0; i < mat.length; i++) {
            List<Integer> result = this.traverse1(i, mat);
            if(reverse == true) {
                this.reverse(result);
            }
            for(int num: result) {
                answer[index++] = num;
            }
            reverse = !reverse;
        }
        for(int j = 1; j < mat[0].length; j++) {
            List<Integer> result = this.traverse2(j, mat);
            if(reverse == true) {
                this.reverse(result);
            }
            for(int num: result) {
                answer[index++] = num;
            }
            reverse = !reverse;
        }
        return answer;
    }
}

File: LEETCODE_498_Diagonal_Traverse.cpp
#include<bits/stdc++.h>
using namespace std;
class Solution {
    vector<int> answer;
    vector<int> traverse1(int& row, vector<vector<int>>& mat) {
        vector<int> result;
        for(int i = row, j = 0; i >= 0 and j < mat[0].size(); i--, j++) {
            result.push_back(mat[i][j]);
        }
        return result;
    }
    vector<int> traverse2(int& col, vector<vector<int>>& mat) {
        vector<int> result;
        for(int i = mat.size()-1, j = col; i >= 0 and j < mat[0].size() ; i--, j++) {
            result.push_back(mat[i][j]);
        }
        return result;
    }
public:
    vector<int> findDiagonalOrder(vector<vector<int>>& mat) {
        bool Reverse = false;
        for(int i = 0; i < mat.size(); i++) {
            vector<int> result = this->traverse1(i, mat);
            if(Reverse == true) {
                reverse(result.begin(), result.end());
            }
            for(int& num: result) {
                this->answer.push_back(num);
            }
            Reverse = !Reverse;
        }
        for(int j = 1; j < mat[0].size(); j++) {
            vector<int> result = this->traverse2(j, mat);
            if(Reverse == true) {
                reverse(result.begin(), result.end());
            }
            for(int& num: result) {
                this->answer.push_back(num);
            } 
            Reverse = !Reverse;
        }
        return this->answer;
    }
};